app/parse.py

```diff
@@ -15,8 +39,102 @@ class Product:
     num_of_reviews: int
 
 
+def parse_single_product(product_soup: Tag) -> Product:
+    return Product(
+        title=product_soup.select_one(".title")["title"],
+        description=product_soup.select_one(".description").text,
+        price=float(product_soup.select_one(".price").text.replace("$", "")),
+        rating=int(product_soup.select_one("p[data-rating]")["data-rating"]),
+        num_of_reviews=int(
+            product_soup.select_one("p.review-count").text.split()[0]
+        ),
+    )
+
+
+def save_to_csv(products: list[Product], file_name: str) -> None:
+    with open(file_name, "w", newline='') as f:
+        writer = csv.writer(f)
+        writer.writerow(FIELDS)
+        writer.writerows([astuple(product) for product in products])
+
+
+def parse_single_product_from_detail_page(url: str) -> Product:
+    tablet_url = urljoin(BASE_URL, url)
+    tablet_page = requests.get(tablet_url).content
+    tablet_soup = BeautifulSoup(tablet_page, "html.parser")
+
+    return Product(
+        title=tablet_soup.select_one(".title").text,
+        description=tablet_soup.select_one(".description").text,
+        price=float(tablet_soup.select_one(".price").text.replace("$", "")),
+        rating=len(tablet_soup.select("p.review-count span")),
+        num_of_reviews=int(
+            tablet_soup.select_one("p.review-count").text.split()[0]
+        ),
+    )
+
+
+def get_all_products_from_their_detail_page(main_url: str) -> list[Product]:
+    with webdriver.Chrome(service=service, options=chrome_options) as driver:
+        driver.get(main_url)
+
+        previous_product_count = 0
+        while True:
+            try:
+                soup = BeautifulSoup(driver.page_source, "html.parser")
+                products_on_page = len(soup.select(".thumbnail"))
+
+                if products_on_page == previous_product_count:
+                    print("Downloaded all products")
+                    break
+
+                previous_product_count = products_on_page
+
+                more_button = WebDriverWait(driver, 2).until(
+                    EC.element_to_be_clickable((By.CLASS_NAME, "ecomerce-items-scroll-more"))
+                )
+                more_button.click()
+
+            except TimeoutException:
+                print("Кнопка 'More' не знайдена або не клікабельна.")
```

Comment:
The error message is in Ukrainian. Ensure that all log messages are in a consistent language, preferably English, to maintain consistency throughout the codebase.